src/core/ext/filters/http/message_compress/message_compress_filter.cc

```diff
@@ -93,130 +103,102 @@ struct call_data {
-      char* val = grpc_slice_to_c_string(GRPC_MDVALUE(md));
-      gpr_log(GPR_ERROR,
-              "Invalid compression algorithm: '%s' (previously disabled). "
-              "Ignoring.",
-              val);
-      gpr_free(val);
-      calld->message_compression_algorithm = GRPC_MESSAGE_COMPRESS_NONE;
-      stream_compression_algorithm = GRPC_STREAM_COMPRESS_NONE;
-    }
-    *has_compression_algorithm = true;
+    GPR_ASSERT(grpc_compression_algorithm_parse(GRPC_MDVALUE(md),
+                                                &compression_algorithm));
+    // Remove this metadata since it's an internal one (i.e., it won't be
+    // transmitted out).
     grpc_metadata_batch_remove(
         initial_metadata,
         initial_metadata->idx.named.grpc_internal_encoding_request);
-    calld->message_compression_algorithm =
-        grpc_compression_algorithm_to_message_compression_algorithm(
-            compression_algorithm);
-    stream_compression_algorithm =
-        grpc_compression_algorithm_to_stream_compression_algorithm(
-            compression_algorithm);
-  } else {
-    /* If no algorithm was found in the metadata and we aren't
-     * exceptionally skipping compression, fall back to the channel
-     * default */
-    if (channeld->default_compression_algorithm != GRPC_COMPRESS_NONE) {
-      calld->message_compression_algorithm =
-          grpc_compression_algorithm_to_message_compression_algorithm(
-              channeld->default_compression_algorithm);
-      stream_compression_algorithm =
-          grpc_compression_algorithm_to_stream_compression_algorithm(
-              channeld->default_compression_algorithm);
-    }
-    *has_compression_algorithm = true;
   }
+  // Check if that algorithm is enabled. Note that GRPC_COMPRESS_NONE is always
+  // enabled.
+  // TODO(juanlishen): Maybe use channel default or abort() if the algorithm
+  // from the initial metadata is disabled.
+  if (GPR_LIKELY(GPR_BITGET(channeld->enabled_compression_algorithms_bitset,
+                            compression_algorithm))) {
+    return compression_algorithm;
+  }
+  const char* algorithm_name;
+  grpc_compression_algorithm_name(compression_algorithm, &algorithm_name);
+  gpr_log(GPR_ERROR,
+          "Invalid compression algorithm: '%s' (previously disabled). "
+          "Will not compress.",
+          algorithm_name);
+  return GRPC_COMPRESS_NONE;
+}
 
+static grpc_error* process_send_initial_metadata(
+    grpc_call_element* elem,
+    grpc_metadata_batch* initial_metadata) GRPC_MUST_USE_RESULT;
+static grpc_error* process_send_initial_metadata(
+    grpc_call_element* elem, grpc_metadata_batch* initial_metadata) {
+  call_data* calld = static_cast<call_data*>(elem->call_data);
+  channel_data* channeld = static_cast<channel_data*>(elem->channel_data);
+  // Find the compression algorithm.
+  grpc_compression_algorithm compression_algorithm =
+      find_compression_algorithm(initial_metadata, channeld);
```

Comment:
So `find_compression_algorithm` is actually just finding what message compression algorithm to use. I wonder if this can be made more verbose since it confused me a little bit.